lib/smart_proxy_remote_execution_ssh/runners/script_runner.rb

```diff
@@ -258,38 +226,61 @@ def indent_multiline(string)
-      ssh_options[:password] = @ssh_password if @ssh_password
-      ssh_options[:passphrase] = @key_passphrase if @key_passphrase
-      ssh_options[:keys_only] = true
-      # if the host public key is contained in the known_hosts_file,
-      # verify it, otherwise, if missing, import it and continue
-      ssh_options[:verify_host_key] = VERIFY_HOST_KEY
-      ssh_options[:auth_methods] = available_authentication_methods
-      ssh_options[:user_known_hosts_file] = prepare_known_hosts if @host_public_key
-      ssh_options[:number_of_password_prompts] = 1
-      ssh_options[:verbose] = settings[:ssh_log_level]
-      ssh_options[:logger] = Proxy::RemoteExecution::Ssh::LogFilter.new(Proxy::Dynflow::Log.instance)
-      return ssh_options
+      @session && @cleanup_working_dirs
+    end
+
+    # Creates session with three pipes - one for reading and two for
+    # writing. Similar to `Open3.popen3` method but without creating
+    # a separate thread to monitor it.
+    def session(args, in_stream: true, out_stream: true, err_stream: true)
+      @session = true
+
+      in_read, in_write = in_stream ? IO.pipe : '/dev/null'
+      out_read, out_write = out_stream ? IO.pipe : [nil, '/dev/null']
+      err_read, err_write = err_stream ? IO.pipe : [nil, '/dev/null']
+      command_pid = spawn(*args, :in => in_read, :out => out_write, :err => err_write, :close_others => false)
+      in_read.close if in_stream
+      out_write.close if out_stream
+      err_write.close if err_stream
+
+      return command_pid, in_write, out_read, err_read
+    end
+
+    def ssh_options(with_pty = false)
+      ssh_options = []
+      ssh_options << "-tt" if with_pty
+      ssh_options << "-o User=#{@ssh_user}"
+      ssh_options << "-o Port=#{@ssh_port}" if @ssh_port
+      ssh_options << "-o IdentityFile=#{@client_private_key_file}" if @client_private_key_file
+      ssh_options << "-o IdentitiesOnly=yes"
+      ssh_options << "-o StrictHostKeyChecking=no"
+      ssh_options << "-o PreferredAuthentications=#{available_authentication_methods.join(',')}"
+      ssh_options << "-o UserKnownHostsFile=#{prepare_known_hosts}" if @host_public_key
+      ssh_options << "-o NumberOfPasswordPrompts=1"
+      ssh_options << "-o LogLevel=#{settings[:ssh_log_level]}"
+      ssh_options << "-o ControlMaster=auto"
+      ssh_options << "-o ControlPath=#{local_command_file("socket")}"
+      ssh_options << "-o ControlPersist=yes"
     end
 
     def settings
       Proxy::RemoteExecution::Ssh::Plugin.settings
     end
 
+    def create_fd(sshpass)
+      io_r, io_w = IO.pipe
+      io_w << sshpass
+      io_w.close
+      io_r
+    end
+
+    def get_args(command, with_pty = false)
+      args = []
+      args += ['/usr/bin/sshpass', '-d', create_fd(@ssh_password).fileno.to_s] if @ssh_password
+      args += ['/usr/bin/sshpass', '-P', 'passphrase', '-d', create_fd(@key_passphrase).fileno.to_s] if @key_passphrase
```

Comment:
In my setup I have a passphrase-less key and a password. When I log in with the key it works, when I log in with the password it works, when I try to log in with the key and a passphrase it works, although the passphrase is never really used.

However when I provide both a passphrase and a password, it gets stuck when opening the initial connection.

```
# pstree -p 2688
sshpass(2688)───sshpass(2689)───ssh(2690)

# ps -f -p 2688 -p 2689 -p 2690 | cat
UID        PID  PPID  C STIME TTY          TIME CMD
foreman+  2688  2531  0 11:54 ?        00:00:00 /usr/bin/sshpass -d 16 /usr/bin/sshpass -P passphrase -d 18 /usr/bin/ssh skilled-foal.example.com -o User=root -o Port=22 -o IdentityFile=/var/lib/foreman-proxy/ssh/id_rsa_foreman_proxy -o IdentitiesOnly=yes -o StrictHostKeyChecking=no -o PreferredAuthentications=password,publickey -o NumberOfPasswordPrompts=1 -o LogLevel=fatal -o ControlMaster=auto -o ControlPath=/var/tmp/foreman-proxy/foreman-ssh-cmd-cea5225c-d810-4349-9c63-38616a32d692/socket -o ControlPersist=yes mkdir -p /var/tmp/foreman-ssh-cmd-cea5225c-d810-4349-9c63-38616a32d692
foreman+  2689  2688  0 11:54 pts/2    00:00:00 /usr/bin/sshpass -P passphrase -d 18 /usr/bin/ssh skilled-foal.example.com -o User=root -o Port=22 -o IdentityFile=/var/lib/foreman-proxy/ssh/id_rsa_foreman_proxy -o IdentitiesOnly=yes -o StrictHostKeyChecking=no -o PreferredAuthentications=password,publickey -o NumberOfPasswordPrompts=1 -o LogLevel=fatal -o ControlMaster=auto -o ControlPath=/var/tmp/foreman-proxy/foreman-ssh-cmd-cea5225c-d810-4349-9c63-38616a32d692/socket -o ControlPersist=yes mkdir -p /var/tmp/foreman-ssh-cmd-cea5225c-d810-4349-9c63-38616a32d692
foreman+  2690  2689  0 11:54 pts/3    00:00:00 /usr/bin/ssh skilled-foal.example.com -o User=root -o Port=22 -o IdentityFile=/var/lib/foreman-proxy/ssh/id_rsa_foreman_proxy -o IdentitiesOnly=yes -o StrictHostKeyChecking=no -o PreferredAuthentications=password,publickey -o NumberOfPasswordPrompts=1 -o LogLevel=fatal -o ControlMaster=auto -o ControlPath=/var/tmp/foreman-proxy/foreman-ssh-cmd-cea5225c-d810-4349-9c63-38616a32d692/socket -o ControlPersist=yes mkdir -p /var/tmp/foreman-ssh-cmd-cea5225c-d810-4349-9c63-38616a32d692

# gdb -p 2688
(gdb) info stack
#0  0x00007effc1de5bbb in pselect () from /lib64/libc.so.6
#1  0x00000000004018f2 in runprogram ()
#2  0x00007effc1d12555 in __libc_start_main () from /lib64/libc.so.6
#3  0x000000000040114d in _start ()

# gdb -p 2689
(gdb) info stack
#0  0x00007f7798a17bbb in pselect () from /lib64/libc.so.6
#1  0x00000000004018f2 in runprogram ()
#2  0x00007f7798944555 in __libc_start_main () from /lib64/libc.so.6
#3  0x000000000040114d in _start ()

# gdb -p 2690
(gdb) info stack
#0  0x00007fb519e24b40 in __read_nocancel () from /lib64/libc.so.6
#1  0x0000555de953b504 in readpassphrase ()
#2  0x0000555de9512eb8 in read_passphrase ()
#3  0x0000555de94e916d in userauth_passwd ()
#4  0x0000555de94ec13b in userauth ()
#5  0x0000555de94ec271 in input_userauth_failure ()
#6  0x0000555de9516ce9 in ssh_dispatch_run ()
#7  0x0000555de94ebb25 in ssh_userauth2 ()
#8  0x0000555de94e69bf in ssh_login ()
#9  0x0000555de94d5bae in main ()
```